File: 678-valid-parenthesis-string/valid-parenthesis-string.php
class Solution {

    /**
     * @param String $s
     * @return Boolean
     */
    function checkValidString($s) {
        $min = 0;
        $max = 0;
        $length = strlen($s); // Get the length of the string

        for ($i = 0; $i < $length; $i++) {
            if ($s[$i] == '(') {
                $min += 1;
                $max += 1;
            } else if ($s[$i] == ')') {
                $min -= 1;
                $max -= 1;
            } else { // $s[$i] == '*'
                $min -= 1;
                $max += 1;
            }

            if ($min < 0) {
                $min = 0;
            }

            if ($max < 0) {
                return false;
            }
        }

        return $min == 0;
    }
}

File: 678-valid-parenthesis-string/valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        min_o=0
        max_o=0
        for i in s:
            if i == '(':
                min_o+=1
                max_o+=1
            elif i ==  ')':
                min_o-=1
                max_o-=1
            else:
                min_o-=1
                max_o+=1
            if min_o < 0:
                min_o = 0
            if max_o <0:
                return False
        return min_o == 0
